src/renderToString.ts

```diff
@@ -0,0 +1,217 @@
+
+// NOTE this should come from `vue`
+type PublicProps = VNodeProps & AllowedComponentProps & ComponentCustomProps
+
+type ComponentMountingOptions<T> = T extends DefineComponent<
+  infer PropsOrPropOptions,
+  any,
+  infer D,
+  any,
+  any
+>
+  ? MountingOptions<
+      Partial<ExtractDefaultPropTypes<PropsOrPropOptions>> &
+        Omit<
+          Readonly<ExtractPropTypes<PropsOrPropOptions>> & PublicProps,
+          keyof ExtractDefaultPropTypes<PropsOrPropOptions>
+        >,
+      D
+    > &
+      Record<string, any>
+  : MountingOptions<any>
+
+// Class component (without vue-class-component) - no props
+export function renderToString<V extends {}>(
+  originalComponent: {
+    new (...args: any[]): V
+    __vccOpts: any
+  },
+  options?: MountingOptions<any> & Record<string, any>
+): Promise<string>
+
+// Class component (without vue-class-component) - props
+export function renderToString<V extends {}, P>(
+  originalComponent: {
+    new (...args: any[]): V
+    __vccOpts: any
+    defaultProps?: Record<string, Prop<any>> | string[]
+  },
+  options?: MountingOptions<P & PublicProps> & Record<string, any>
+): Promise<string>
+
+// Class component - no props
+export function renderToString<V extends {}>(
+  originalComponent: {
+    new (...args: any[]): V
+    registerHooks(keys: string[]): void
+  },
+  options?: MountingOptions<any> & Record<string, any>
+): Promise<string>
+
+// Class component - props
+export function renderToString<V extends {}, P>(
+  originalComponent: {
+    new (...args: any[]): V
+    props(Props: P): any
+    registerHooks(keys: string[]): void
+  },
+  options?: MountingOptions<P & PublicProps> & Record<string, any>
+): Promise<string>
+
+// Functional component with emits
+export function renderToString<Props extends {}, E extends EmitsOptions = {}>(
+  originalComponent: FunctionalComponent<Props, E>,
+  options?: MountingOptions<Props & PublicProps> & Record<string, any>
+): Promise<string>
+
+// Component declared with defineComponent
+export function renderToString<
+  PropsOrPropOptions = {},
+  RawBindings = {},
+  D = {},
+  C extends ComputedOptions = ComputedOptions,
+  M extends MethodOptions = MethodOptions,
+  Mixin extends ComponentOptionsMixin = ComponentOptionsMixin,
+  Extends extends ComponentOptionsMixin = ComponentOptionsMixin,
+  E extends EmitsOptions = Record<string, any>,
+  EE extends string = string,
+  PP = PublicProps,
+  Props = Readonly<ExtractPropTypes<PropsOrPropOptions>>,
+  Defaults extends {} = ExtractDefaultPropTypes<PropsOrPropOptions>
+>(
+  component: DefineComponent<
+    PropsOrPropOptions,
+    RawBindings,
+    D,
+    C,
+    M,
+    Mixin,
+    Extends,
+    E,
+    EE,
+    PP,
+    Props,
+    Defaults
+  >,
+  options?: MountingOptions<
+    Partial<Defaults> & Omit<Props & PublicProps, keyof Defaults>,
+    D
+  > &
+    Record<string, any>
+): Promise<string>
+
+// component declared by vue-tsc ScriptSetup
```

Comment:
Did we copy paste this entire thing? Is it possible to share it somehow?

Reply:
@lmiller1990 Yes, it's nearly the same as mount, but returns a `Promise<string>`. I tried to extract it with a mapped type, but sadly all the tricks I found did not work (you can map one signature, but the overloads aren't mapped)😞